### PR TITLE
If Branch Constant Folding

### PR DESCRIPTION
**Pull request by yuslepukhin**
### Description

When and if `If` condition proves to be a constant value, inline the corresponding subgraph yielding to more constant folding and optimization.

### Motivation and Context
Newly converted models feature lots of nested `If` nodes that can be inlined and collapsed.

In particular, for the sample models we are gaining on TorchScript exported models.
For `HF Mobile Bert Dynamo` runtime went down from 0.069 -> 0.046. In total, AOT inlining + `If` constant folding
yields improvement of about 50% 0.102 -> 0.046. Brining us very close to TorchScript exported models.

`HF Bart Dynamo` further improves 0.668 -> 0.45. AOT + `If` constant folding improves 0.98 -> 0.45

Earlier the size of 
HF Mobile Bert **161Mb+**, now **98Mb**
HF Bart Dynamo pre-optimized model was about **1.2Gb**. It is now **710MB**

image

### CODE REVIEW

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 1]
```diff
@@ -118,7 +183,22 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
     }
 
     bool converted_to_constant = false;
-    if (node->OpType().compare("Shape") == 0) {
+    if (node->OpType().compare("If") == 0) {
+      // This process constant folds the If node only,
+      // but inlines the nodes of the corresponding branch graph.
+      // It does not convert the node to a constant in a common sense.
+      // We call it constant folding because the `If` node constant condition
+      // may enable use to inline the corresponding branch graph.
```
Comment (yuslepukhin) [inline onnxruntime/core/optimizer/constant_folding.cc:191]: `use` -> `us`

File: include/onnxruntime/core/graph/graph.h [hunk 2]
```diff
@@ -982,6 +982,16 @@ class Graph {
   */
   bool RemoveNode(NodeIndex node_index);
 
+  /** The function regenerates NodeProto list in a topological order of the Graph
+  nodes instance and replaces it in the graph_proto_. This is necessary for subgraphs
+  that were subjected to constnat folding and some nodes were removed from Graph, but
```
Comment (yuslepukhin) [inline include/onnxruntime/core/graph/graph.h:987]: constant

File: onnxruntime/core/graph/graph.cc [hunk 3]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+      std::string new_name;
+      auto hit = name_mapping.find(constant_name);
+      if (hit != name_mapping.cend()) {
+        new_name = hit->second;
+      } else {
+        new_name = GenerateNodeArgName(make_unique(constant_name));
+        name_mapping.emplace(constant_name, new_name);
+      }
+      ORT_RETURN_IF_ERROR(AddConstantProtoAsInitializer(constant_node_proto, new_name));
+    }
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4167]: IIRC the onnxruntime::Graph ctor converts all Constant nodes to initializers, so given this is a Graph instance I wouldn't have expected any Constant nodes to remain. Different story if you were inlining the GraphProto in the then/else attribute though.

File: onnxruntime/core/graph/graph.cc [hunk 4]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+  }
+
+  for (const auto& init : graph_to_inline.name_to_initial_tensor_) {
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4170]: nit: `init` has multiple meanings and is a little vague in this context. name_initializer_pair would be clearer.

@edgchen1 is there a C++17 way to unpack the two values in the pair into named variables here?

File: onnxruntime/core/graph/graph.cc [hunk 5]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+    const gsl::not_null<TensorProto*> tensor{graph_proto_->add_initializer()};
+    *tensor = *init.second;
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4172]: is it possible to std::move the initializer? if the initializer is large this could potentially impact peak memory usage.

File: onnxruntime/core/graph/graph.cc [hunk 6]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+      const auto& constant_name = constant_node_proto.output(0);
+      // Check if this is an output of the graph
+      std::string new_name;
+      auto hit = name_mapping.find(constant_name);
+      if (hit != name_mapping.cend()) {
+        new_name = hit->second;
+      } else {
+        new_name = GenerateNodeArgName(make_unique(constant_name));
+        name_mapping.emplace(constant_name, new_name);
+      }
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4165]: Can this be refactored into a lambda returning the new name to reduce binary size impact? logic is repeated below

File: onnxruntime/core/graph/graph.cc [hunk 7]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+
+  // We want to make sure we get nodes in topological order
+  // because Constant folding may cause the nodes appear in
+  // a different order.
+  GraphViewer graph(graph_to_inline);
+  for (const auto node_idx : graph.GetNodesInTopologicalOrder()) {
+    const auto* node = graph_to_inline.GetNode(node_idx);
+    if (node == nullptr || node->OpType() == kConstant) {
+      continue;
+    }
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4209]: wouldn't have expected either of these to be possible. IIRC the graph viewer filters out removed nodes, and any Constant was converted to an initializer in the ctor of the underlying Graph instance.

File: onnxruntime/core/graph/graph.cc [hunk 8]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+
+    InlinedVector<NodeArg*> new_node_input_defs;
+    for (const auto* input_def : node->InputDefs()) {
+      // Check if this is one of the implicit graph inputs
+      // then leave the name as is and reuse the NodeArg
+      const auto& input_name = input_def->Name();
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4215]: check Exists() first to handle missing optional inputs.

File: onnxruntime/core/graph/graph.cc [hunk 9]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+      if (if_all_inputs.count(input_name) > 0) {
+        auto* node_arg = GetNodeArgIncludingParentGraphs(input_name);
+        ORT_ENFORCE(node_arg != nullptr,
+                    "Implicit or explicit input is expected to have existing node_arg: ", input_name);
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4219]: nit: use an `assert` for this. there's lots of testing to validate all values have a NodeArg in all applicable subgraphs and I don't think this shouldn't be the place enforcing that. binary size cost is the main reason for not doing it here.

File: onnxruntime/core/graph/graph.cc [hunk 10]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+    const auto new_node_name = GenerateNodeName(make_unique(node->Name()));
+    if (node->ContainsSubgraph()) {
+      // Make a copy
+      auto attributes = node->GetAttributes();
+      RenameNodeAttributesSubgraphDependentNames(name_mapping, attributes);
+      AddNode(new_node_name, node->OpType(), node->Description(),
+              new_node_input_defs,
+              new_node_output_defs,
+              &attributes,
+              node->Domain());
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:4263]: Feels like it adds a lot of complication to go back to updating protobuf based types when all the subgraphs would have onnxruntime::Graph instances already, and we could recurse into those to do naming updates more easily. 

Can we instead 
- recurse into the subgraphs returned by Node::GetAttributeNameToMutableSubgraphMap() (need to use that to get mutable Graph instances) and update names directly
  - we only need to recurse if an implicit input of the node containing the subgraph/s is present in the renaming map

If we do that, I don't think the node attributes need to be updated at all (a std::move would be ideal to avoid the copy cost)  The only time the GraphProto attributes matter is if we're serializing the updated model via Graph::ToGraphProto, and there's existing logic in Node::ToProto to update the subgraph attributes based on their Graph instances (`update_subgraphs` param) when that happens.

File: onnxruntime/core/graph/graph.cc (hunk 4 above)
Comment (edgchen1) [reply onnxruntime/core/graph/graph.cc:4170]: why yes, there is.
https://en.cppreference.com/w/cpp/language/structured_binding

e.g., `for (const auto& [name, initializer] : graph_to_inline.name_to_initial_tensor_) {`

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 11]
```diff
@@ -90,6 +90,71 @@ static bool ConstantFoldShapeNode(Graph& graph, Node& node) {
...
+
+  // This is a boolean initializer with a single element.
+  const auto proto_shape = utils::GetTensorShapeFromTensorProto(*initializer);
+  // This must be pre-allocated
+  Tensor tensor{
+      DataTypeImpl::TensorTypeFromONNXEnum(initializer->data_type())->GetElementType(),
+      proto_shape,
+      std::make_shared<CPUAllocator>()};
+
+  Status status = utils::TensorProtoToTensor(Env::Default(), graph.ModelPath().ToPathString().c_str(), *initializer, tensor);
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/optimizer/constant_folding.cc:116]: I will see what I can do. I lint everything before pushing changes, otherwise CI fails.

File: onnxruntime/core/graph/graph.cc (hunk 6 above)
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:4165]: I removed the constant nodes loop. lambdas are often inlined though.

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 12]
```diff
@@ -1021,6 +1023,126 @@ TEST_F(GraphTransformationTests, ConstantFoldingStringInitializer) {
...
+  ASSERT_STATUS_OK(session_object.Load(sstr));
+  ASSERT_STATUS_OK(session_object.Initialize());
+
+  // const auto resulting_model_proto = session_object.GetModel().ToProto();
+  // std::string printed_model = ONNX_NAMESPACE::ProtoToString(resulting_model_proto);
+  // ASSERT_FALSE(printed_model.empty());
+  // std::cout << printed_model << std::endl;
+
+  // This is the resulting model proto.
+  // The remaining If node is not constant foldable because Size() does not constant fold
```
Comment (gramalingam) [inline onnxruntime/test/optimizer/graph_transform_test.cc:1121]: This is worth doing (based on real examples seen coming from the dynamo exporter). Is the plan to do it in a separate PR? There is a similar special handling for [Shape operator](https://github.com/microsoft/onnxruntime/blob/dfafcb58aa5fb262b4ff964ccdcc3271672032c0/onnxruntime/core/optimizer/constant_folding.cc#L30) already, and I guess we can do something similar for Size.

File: onnxruntime/core/optimizer/constant_folding.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/optimizer/constant_folding.cc:116]: Linting doesn't (currently) enforce the 120 char limit. You need to manually do that - which is why having a vertical ruler in your editor makes it way easier to do when editing the code. clang-format does a _horrible_ job if you let it it will throw away all rational formatting and try and cram as much as possible on every line which completely screws up readability.

I do have a script to apply clang-format only to long lines that could maybe be plugged in to provide automatic suggestions.

[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/optimizer/constant_folding.cc:116]: Just for you: https://github.com/microsoft/onnxruntime/pull/18293

File: onnxruntime/core/graph/graph.cc (hunk 6 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph.cc:4165]: I would hope in a MinSizeRel build it would not inline. In a Release build I wouldn't care if it did.

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 12 above)
Comment (yuslepukhin) [reply onnxruntime/test/optimizer/graph_transform_test.cc:1121]: Yes, there is a couple of things on my list for separate PR. `Size` and two consecutive `Reshape` nodes for separate PRs.

File: onnxruntime/core/graph/graph.cc [hunk 13]
```diff
@@ -984,6 +945,52 @@ bool Node::ClearAttribute(const std::string& attr_name) {
   graph_->SetGraphProtoSyncNeeded();
   return attributes_.erase(attr_name) > 0;
 }
+
+void Node::ToProto(NodeProto& proto, bool
```
Comment (gramalingam) [inline onnxruntime/core/graph/graph.cc:949]: Just wondering why github is showing the whole function body as changed, when it is only a few lines that have changed. Oh, well.

File: onnxruntime/core/graph/graph.cc [hunk 14]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+    for (auto& input : *node.mutable_input()) {
+      auto hit = name_mapping.find(input);
+      if (hit != name_mapping.cend()) {
+        input = hit->second;
+      }
+    }
+  }
+}
+
+static void RenameNodeAttributesSubgraphDependentNames(const InlinedHashMap<std::string, std::string>& name_mapping,
```
Comment (gramalingam) [inline onnxruntime/core/graph/graph.cc:4095]: I wonder if either this can be dropped or if the duplicated lines in the previous function (lines 4076 onwards) be replaced by a call to this function?

File: onnxruntime/core/graph/graph.cc [hunk 15]
```diff
@@ -4046,6 +4070,210 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+  unique_id = GenerateNodeName(unique_id);
+
+  auto make_unique = [&unique_id](const std::string& name) {
+    return unique_id + '_' + name;
+  };
+
+  // Check if the name is an input or implicit input.
+  // These are not renamed.
+  InlinedHashSet<std::string_view> if_all_inputs;
+  const auto if_inputs = if_node.InputDefs();
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc:4122]: Is either (explicit or implicit) inputs ever used down below? Not sure why we would need that, since the renaming is done only for outputs assigned to in some node within the subgraph.

File: onnxruntime/core/graph/graph.cc (hunk 14 above)
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:4095]: This is being re-worked.

File: onnxruntime/core/graph/graph.cc (hunk 13 above)
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:949]: It was moved.

File: onnxruntime/core/graph/graph.cc [hunk 16]
```diff
@@ -4046,6 +4047,229 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+                                         Graph& graph) {
+  for (auto& node : graph.Nodes()) {
+    if (node.ContainsSubgraph()) {
+      for (auto& [name, subgraph] : node.GetAttributeNameToMutableSubgraphMap()) {
+        RenameSubgraphDependentNames(name_mapping, *subgraph);
+      }
+    }
+
+    // NodeArgs need to be updated
+    auto input_defs = node.MutableInputDefs();
```
Comment (edgchen1) [inline onnxruntime/core/graph/graph.cc:4060]: `input_defs` unused?

File: onnxruntime/core/graph/graph.cc [hunk 17]
```diff
@@ -4046,6 +4047,229 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+    for (auto& input_def : node.MutableInputDefs()) {
+      auto hit = name_mapping.find(input_def->Name());
+      if (hit != name_mapping.cend()) {
+        input_def = graph.GetNodeArgIncludingParentGraphs(hit->second);
```
Comment (edgchen1) [inline onnxruntime/core/graph/graph.cc:4064]: regarding the function name, is it reassigning input node args instead of "renaming"?

File: onnxruntime/core/graph/graph.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc]: The nodes in the graph refer to node inputs, out of scope inputs or the outputs of previously inlined nodes, and the latter  will be new names.

File: onnxruntime/core/graph/graph.cc (hunk 17 above)
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc:4064]: +1. This is a bit unclear to me also.

File: onnxruntime/core/graph/graph.cc [hunk 18]
```diff
@@ -4046,6 +4047,229 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+    if (node.ContainsSubgraph()) {
+      for (auto& [name, subgraph] : node.GetAttributeNameToMutableSubgraphMap()) {
+        RenameSubgraphDependentNames(name_mapping, *subgraph);
+      }
+    }
+
+    // NodeArgs need to be updated
+    auto input_defs = node.MutableInputDefs();
+    for (auto& input_def : node.MutableInputDefs()) {
+      auto hit = name_mapping.find(input_def->Name());
```
Comment (gramalingam) [inline onnxruntime/core/graph/graph.cc:4062]: Should if we check if the InputExists? Or, are you relying on the map not containing an entry for an empty string name?

File: onnxruntime/core/graph/graph.cc [hunk 19]
```diff
@@ -4046,6 +4047,229 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+    if (hit != name_mapping.cend()) {
+      tensor->set_name(hit->second);
+    } else {
+      auto new_name = GenerateNodeArgName(make_unique(src_name));
+      ORT_IGNORE_RETURN_VALUE(name_mapping.emplace(src_name, new_name));
+      tensor->set_name(std::move(new_name));
+    }
+
+    auto insert_result = name_to_initial_tensor_.emplace(tensor->name(), tensor);
+    ORT_ENFORCE(insert_result.second, "Initializer name: ", tensor->name(), " from graph: ",
```
Comment (gramalingam) [inline onnxruntime/core/graph/graph.cc:4160]: It seems to me that we could have a name-conflict here that has nothing to do with specializing code (which, I guess, is involved only with function-inlining). Specifically, what happens if we inline two if-statements one after another? And if the inlined-branches in the two if-statements have initializers with the same name? They don't conflict with each other before if-branch-folding. First if-branch-folding should be fine. But the second one might produce a name conflict, right?

File: onnxruntime/core/graph/graph.cc (hunk 17 above)
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc:4064]: Ok, I guess NodeArgs are unique, and only a pointer is stored, and this is just updating the pointer. So, a minor nit is: why can't the name-mapping actually be a mapping from name to NodeArg-pointer? It looks like in all places, you just need the corresponding NodeArg pointer.

Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:4064]: The node args for the new name are already created with the correct names, when we get here. So we simply query them and replace. Am I missing something?

File: onnxruntime/core/graph/graph.cc (hunk 19 above)
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc:4160]: My fault, I missed the renaming that happens just a few lines above! I was misled by the error message "Check specializing code.", which may be misleading. I guess as long as `make_unqiue` above does its job, this shouldn't fail.

File: onnxruntime/core/graph/graph.cc (hunk 17 above)
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:4064]: I am renaming the function. I am worried that there is no specific proposals as to the name.

File: onnxruntime/core/graph/graph.cc (hunk 19 above)
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:4160]: I have adjusted the error message.

**Review by gramalingam — DISMISSED** [review]
LGTM, thanks!

File: onnxruntime/core/graph/graph.cc (hunk 17 above)
Comment (edgchen1) [reply onnxruntime/core/graph/graph.cc:4064]: > why can't the name-mapping actually be a mapping from name to NodeArg-pointer? It looks like in all places, you just need the corresponding NodeArg pointer.

or even a `map<NodeArg*, NodeArg*>`?

File: include/onnxruntime/core/graph/graph.h [hunk 20]
```diff
@@ -441,6 +443,13 @@ class Node {
     return attr_to_subgraph_map_;
   }
 
+  /** Gets a map of attribute name to the mutable Graph instances for all subgraphs of the Node.
+   * @returns a mutable map of mutable subgraphs.
+   */
+  std::unordered_map<std::string, gsl::not_null<Graph*>>& GetMutableAttributeNameToSubgraphMap() {
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply include/onnxruntime/core/graph/graph.h:449]: I think it should be `GetMutableMapAttributeNameToSubgraph`. This is because keys are not mutable in maps and values are by default IF the map is mutable.

File: onnxruntime/core/graph/graph.cc (hunk 17 above)
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:4064]: Per offline conversation, Name -> NodeArg* should do it.

File: include/onnxruntime/core/graph/graph.h (hunk 20 above)
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply include/onnxruntime/core/graph/graph.h:449]: Per offline discussion: `GetMutableMapOfAttributeNameToSubgraph`

**Review by gramalingam — DISMISSED** [review]

**Review by edgchen1 — DISMISSED** [review]

File: onnxruntime/core/graph/graph.cc [hunk 21]
```diff
@@ -4047,6 +4048,311 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+    for (auto* node : new_nodes) {
+      int arg_pos = -1;
+      for (auto* input_def : node->InputDefs()) {
+        ++arg_pos;
+        auto hit = output_args.find(input_def->Name());
+        if (hit != output_args.cend()) {
+          // The input to this node is an output from a previous node in this graph.
+          // Create relationship between this node (node), and the node providing the output (output_node).
+          const auto& producer = hit->second.first;
+          const auto src_idx = hit->second.second;
```
Comment (edgchen1) [inline onnxruntime/core/graph/graph.cc:4328]:
```suggestion
          // Create relationship between this node (node), and the node providing the output (producer).
          const auto& [producer, src_idx] = hit->second;
```

File: onnxruntime/core/graph/graph.cc [hunk 22]
```diff
@@ -4047,6 +4048,311 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+      // (outputs of If node)
+      arg_pos = -1;
+      for (auto& output_def : node->OutputDefs()) {
+        ++arg_pos;
+        auto hit = input_args.find(output_def->Name());
+        if (hit != input_args.cend()) {
+          // The output of this node is an input to another node in this graph.
+          // Create relationship between this node (node), and the node using the input (input_node).
+          const auto& consumer = hit->second.first;
+          const auto dst_idx = hit->second.second;
```
Comment (edgchen1) [inline onnxruntime/core/graph/graph.cc:4343]:
```suggestion
          // Create relationship between this node (node), and the node using the input (consumer).
          const auto& [consumer, dst_idx] = hit->second;
```

File: onnxruntime/test/testdata/transform/transform_nested_ifs_toplogical_sorted_nodes.txt
Comment (edgchen1) [inline onnxruntime/test/testdata/transform/transform_nested_ifs_toplogical_sorted_nodes.txt:1]: is this file for generating the test onnx model? maybe it should have a `.py` extension.

File: onnxruntime/test/testdata/transform/transform_nested_ifs_toplogical_sorted_nodes.txt [hunk 23]
```diff
@@ -0,0 +1,859 @@
...
+ort_inputs = {'input_0': array(0.8965, dtype=float16)}
+
+# Set up the inference session
+session_options = ort.SessionOptions()
+session_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_DISABLE_ALL
+onnx_model = onnx.ModelProto()
+google.protobuf.text_format.Parse(onnx_model_text, onnx_model)
+
+# Uncomment this line to save the model to a file for examination
+# onnx.save_model(onnx_model, "test_output_match_opinfo__linalg_vector_norm_cpu_float16.onnx")
```
Comment (edgchen1) [inline onnxruntime/test/testdata/transform/transform_nested_ifs_toplogical_sorted_nodes.txt:852]: this model name doesn't match the newly added test model name - is it supposed to?

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 24]
```diff
@@ -90,6 +91,45 @@ static bool ConstantFoldShapeNode(Graph& graph, Node& node) {
...
+
+  if (!status.IsOK()) {
+    LOGS(logger, WARNING) << "Unable to constant fold. InlineIfSubgraph failed "
+                          << " node '" << if_node.Name() << "': "
+                          << status.ErrorMessage();
+    return status;
+  }
+
+  graph_utils::RemoveNodeOutputEdges(graph, if_node);
+  graph.RemoveNode(if_node.Index());
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/optimizer/constant_folding.cc:127]: `ConstantFolding` transformer has its node removal done at the end of the `Apply` function. I originally kept it in the same file. Then one comment suggested that I should move it inside where it is now, so `it is one` place. So it is already in one place. In addition, `InlineIfSubgraph` has its node argument as a reference, so it would be logical that the node stays alive for the duration of the function.

I believe, revisiting this point actually distracts from the main objective.

File: onnxruntime/test/testdata/transform/transform_nested_ifs_toplogical_sorted_nodes.txt
Comment (yuslepukhin) [reply onnxruntime/test/testdata/transform/transform_nested_ifs_toplogical_sorted_nodes.txt:1]: Not really, it is an onnx script. The problem, though, ONNX script parser has a bug, and it generates an invalid model. This `.txt` is for reference only.

File: onnxruntime/core/graph/graph.cc [hunk 25]
```diff
@@ -4221,30 +4255,46 @@ Status Graph::InlineIfSubgraph(bool condition_value, Node& if_node, const loggin
...
+    const auto new_node_name = GenerateNodeName(make_unique(node->OpType()));
     Node& new_node = AddNode(new_node_name, node->OpType(), node->Description(),
                              new_node_input_defs,
                              new_node_output_defs,
                              nullptr,
                              node->Domain());
 
+    if (!is_this_main_graph) {
+      int arg_pos = -1;
+      for (auto* input_def : new_node_input_defs) {
```
Comment (gramalingam) [inline onnxruntime/core/graph/graph.cc:4267]: Is this intended to be different from `map_defs` above, or can this be replaced by a call to `map_defs`? At least the check for `Exists(nodearg)` seems potentially relevant here.

**Review by gramalingam — APPROVED** [review]

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 26]
```diff
@@ -90,6 +91,45 @@ static bool ConstantFoldShapeNode(Graph& graph, Node& node) {
   return is_concrete_shape;  // convert to constant if this is true
 }
 
+// This function inlines the appropriate subgraph. It does not literally fold it.
+static Status ConstantFoldIfNode(Graph& graph, Node& if_node, const logging::Logger& logger, bool& folded) {
```
Comment (pranavsharma) [inline onnxruntime/core/optimizer/constant_folding.cc:95]: minor nit: place out params after input params (https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs)

File: include/onnxruntime/core/graph/graph.h [hunk 27]
```diff
@@ -1134,6 +1143,26 @@ class Graph {
...
+
+    The function will process any subgraphs in each of the nodes being inlined,
+    and will rename any references to the new names introduced.
+
+    @param condition_value If condition value
+    @param if_node - the node that contains the graph_to_inline. This node is going
+    to be deleted and replaced by the corresponding graph (either then or else)
+    @param logger
+  */
+  Status InlineIfSubgraph(bool condition_value, Node& if_node, const logging::Logger& logger);
```
Comment (pranavsharma) [inline include/onnxruntime/core/graph/graph.h:1164]: minor nit: place out params after input params (https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs)

File: include/onnxruntime/core/graph/graph.h [hunk 28]
```diff
@@ -586,7 +595,7 @@ class Node {
   // create a Graph instance for an attribute that contains a GraphProto
   void CreateSubgraph(const std::string& attr_name);
 
-  const std::vector<std::unique_ptr<Graph>>& MutableSubgraphs() noexcept { return subgraphs_; }
+  std::vector<std::unique_ptr<Graph>>& MutableSubgraphs() noexcept { return subgraphs_; }
```
Comment (pranavsharma) [inline include/onnxruntime/core/graph/graph.h:598]: nit: inconsistent naming. Some of the get functions are named like GetFoo(); some are just Foo().

File: onnxruntime/core/graph/graph.cc [hunk 29]
```diff
@@ -4047,6 +4048,301 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+  // Name mapping from the graph to inline to the graph we are inlining into
+  // we also use this to process any subgraphs in the graph we are inlining
+  InlinedHashMap<std::string, NodeArg*> name_to_nodearg;
+
+  // We are going to map the outputs of the graph to inline to the outputs of the If node.
+  // They are assumed to be in the same order.
+  const auto node_output_defs = if_node.MutableOutputDefs();
+  const auto graph_output_defs = graph_to_inline.GetOutputs();
+  for (size_t i = 0; i < graph_output_defs.size(); ++i) {
+    name_to_nodearg.emplace(graph_output_defs[i]->Name(), node_output_defs[i]);
```
Comment (pranavsharma) [inline onnxruntime/core/graph/graph.cc:4127]: Can we call reserve() on name_to_nodearg before emplacing?

File: onnxruntime/core/graph/graph.cc [hunk 30]
```diff
@@ -4047,6 +4048,301 @@ Status Graph::AddConstantProtoAsInitializer(const ONNX_NAMESPACE::NodeProto& nod
...
+  if (condition_value) {
+    sub_graph = if_node.GetMutableGraphAttribute(then_branch);
+  } else {
+    sub_graph = if_node.GetMutableGraphAttribute(else_branch);
+  }
+
+  if (sub_graph == nullptr) {
+    auto str = MakeString("Unable to constant fold If node: '", if_node.Name(), "' Unable to fetch: ",
+                          (condition_value ? then_branch : else_branch));
+    LOGS(logger, WARNING) << str;
```
Comment (pranavsharma) [inline onnxruntime/core/graph/graph.cc:4085]: Do we want the users to do something when they see this warning? If not, can it be INFO?

**Review by pranavsharma — COMMENTED** [review]
Some minor comments.

**Review by pranavsharma — APPROVED** [review]

File: onnxruntime/core/graph/graph.cc (hunk 29 above)
Comment (yuslepukhin) [reply onnxruntime/core/graph/graph.cc:4127]: No, we can not. This requires traversing all the nodes and NodeArgs in advance.
